src/gc/proxy.d

```diff
@@ -28,32 +28,67 @@ private
     extern (C) void thread_init();
     extern (C) void thread_term();
 
+    import core.internal.spinlock;
+    static SpinLock instanceLock;
+
     __gshared GC instance;
     __gshared GC proxiedGC; // used to iterate roots of Windows DLLs
-
 }
 
+alias GCInitNoThrow = void function() nothrow @nogc;
 
 extern (C)
 {
 
     void gc_init()
     {
-        config.initialize();
-        ManualGC.initialize(instance);
-        ConservativeGC.initialize(instance);
-        if (instance is null)
-        {
-            import core.stdc.stdio : fprintf, stderr;
-            import core.stdc.stdlib : exit;
+        import core.atomic : atomicLoad, atomicStore;
+
+        auto pinstance = cast(shared(GC)*) &instance;
 
-            fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
-            exit(1);
+        if (atomicLoad(*pinstance) is null)
+        {
+            instanceLock.lock();
+            scope(exit) instanceLock.unlock();
+            {
+                if (atomicLoad(*pinstance) is null)
+                {
+                    GC auxInstance;
+                    config.initialize();
+                    ManualGC.initialize(auxInstance);
+                    ConservativeGC.initialize(auxInstance);
+
+                    if (auxInstance is null)
+                    {
+                        import core.stdc.stdio : fprintf, stderr;
+                        import core.stdc.stdlib : exit;
+
+                        fprintf(stderr, "No GC was initialized, please recheck the name of the selected GC ('%.*s').\n", cast(int)config.gc.length, config.gc.ptr);
+                        instanceLock.unlock();
+                        exit(1);
+                    }
+
+                    atomicStore(*pinstance, cast(shared GC)auxInstance);
+
+                    // NOTE: The GC must initialize the thread library
+                    //       before its first collection.
+                    thread_init();
```

Comment:
What do you indent to do about [`Thread.getThis`](https://dlang.org/phobos/core_thread.html#.Thread.getThis)? It is currently a GC allocated instance that is eagerly created in [`thread_init`](https://github.com/dlang/druntime/blob/a66f9fa78f2e295146fb2b3fc8e567983c9e9aed/src/core/thread.d#L2023).

Reply:
That may be why *pinstance had to be set just above! Seems fragile. Perhaps we can not use the GC for getThis?

Reply:
You can currently retain a Thread object longer than the threads lives, e.g. there is even a Thread.join methods, so there isn't a clear owner of that instance. Not too easy to avoid the GC here.